source/binary.cpp

```diff
@@ -590,7 +590,7 @@ spv_result_t Parser::parseOperand(size_t inst_offset,
     case SPV_OPERAND_TYPE_OPTIONAL_LITERAL_STRING: {
       const size_t max_words = _.num_words - _.word_index;
       std::string string =
-          spvtools::utils::MakeString(_.words + _.word_index, max_words, false);
+          spvtools::utils::MakeString(_.native_words.get() + _.word_index, max_words, false);
```

Comment:
I'm not sure this change is correct. The SPEC says:

> A string is interpreted as a nul-terminated stream of characters. All string comparisons are case sensitive. The character set is Unicode in the UTF-8 encoding scheme. The UTF-8 octets (8-bit bytes) are packed four per [word](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Word), following the little-endian convention (i.e., the first octet is in the lowest-order 8 bits of the word).

How do you know that the `native_words` are the correct little-endian order?

Reply:
While the spec confusingly says "little-endian" it actually calls for packing bytes into 4-byte words by shifting them, which is what `MakeString` implements. This means that the `spv\0` (`[73,70,76,0]`) string gets turned into the `0x00767073` word which will end up like `73 70 76 00` on a little-endian machine, but on a big-endian machine it will be `00 76 70 73`. Very much not a byte string that can be read directly!

We therefore want all the words to be in the native order because `MakeString` is actually endinaness-agnostic as it only looks at the 32-bit words, it never tries to cast them to chars and read the result as a string. Only looking at things at the word level means endianness conversion is just pre/post processing to swap the words to match machine order, with zero additional complexity required.